### PR TITLE
feat: add feature flags

### LINKED ISSUE #334: Breaking changes in tx handler and old transactions
We need to deal with breaking changes in transaction checks to avoid the situation when we have old transactions that rejected by the new rules.

Example:

> I was trying to sync local node with testnet for #333 and failed
> 
> > ...
> > but i was not able to sync local node with testnet because of [this tx](https://testnet.leapdao.org/explorer/tx/0x26805fe0b0e1cf7ab4dfede7f19c76493b2ec9b4490a5b42ed467374b30b9b94). I've got not touched 0x197970e48082cd46f277abdb8afe492bccd78300 error at height 43897 in tx handler
> > ...
> 
> It happens here: https://github.com/leapdao/leap-node/blob/master/src/tx/applyTx/checkSpendCond.js#L522
> 
> We have added this rule after some spending conditions transactions landed on testnet.

# Bounty

## Scope

Implement the mechanism of feature flags that should allow to enable certain flags at the defined height

1. We have a set of feature flags (erc1948_new_bytecode, spend_cond_not_touched and so on)
2. Network config with height → flag mapping (at which height this flag should be enabled). If there is no height it enabled by default (for not affected networks)
3. We use these flags in checks (flag value can be derived from current height and)

## Deliverables

1. The code
2. Unit tests
3. Integration tests
4. Docs
5. Feature flag for «not touched» rule in `checkSpendCond` check

## Gain for the project
Eliminates fear of break everything with one wrong move

### Publicly visible delivery

Unlocks https://github.com/leapdao/leapdao-docs/issues/45

## Definition of done

Reviewer should be able to resync local node with testnet from scratch

## Roles
bounty gardener: @sunify / 15%
bounty worker: @sunify / 65%
bounty reviewer: name / 20%

### PR DESCRIPTION
**Pull request by sunify**
Resolves #334

### CODE REVIEW

File: src/flags/flagsFactory.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/* eslint-disable no-prototype-builtins */
+
+function flagValue(bridgeState, flagHeights, flag) {
+  if (!flagHeights.hasOwnProperty(flag)) {
+    return true;
+  }
+
+  const targetHeight = flagHeights[flag];
+  return bridgeState.blockHeight >= targetHeight;
```
Comment (troggy) [inline src/flags/flagsFactory.js]: `bridgeState.blockHeight` is updated only every 32 blocks

https://github.com/leapdao/leap-node/blob/d7280b3ce39fc9cdf9ebee47247442d236743ae4/src/block/index.js#L44

Comment (troggy) [reply src/flags/flagsFactory.js]: if a single 32-blocks chunk has both types of txs, how to make a switch?

**Comment by sunify** [comment]
Config to resync testnet (added only `flagHeights`):

```json
{
  "exitHandlerAddr": "0x26a937302cc6A0A7334B210de06136C8C61BA885",
  "bridgeAddr": "0xC449D4CD1dEc611d8cA5Fd8167Bf46d6e6d345b9",
  "operatorAddr": "0xb3356900d56F39c79Bfdc2b625d15B1b5b9262a9",
  "rootNetworkId": 4,
  "network": "leap-testnet-theta",
  "networkId": 218508104,
  "eventsDelay": 2,
  "bridgeDelay": 6,
  "version": "v5.3.3",
  "genesis": {
    "genesis_time": "2019-05-23T12:23:22.546083464Z",
    "chain_id": "test-chain-H5Ijd9",
    "consensus_params": {
      "block": {
        "max_bytes": "22020096",
        "max_gas": "-1",
        "time_iota_ms": "1000"
      },
      "evidence": {
        "max_age": "100000"
      },
      "validator": {
        "pub_key_types": [
          "ed25519"
        ]
      }
    },
    "validators": [
      {
        "address": "4881A447F30F484F046859530083FD27C5F7A70B",
        "pub_key": {
          "type": "tendermint/PubKeyEd25519",
          "value": "Su7amMRUBc+MinqO026cuI3Q4jinwx0KJRdjXDp4Rao="
        },
        "power": "10",
        "name": ""
      }
    ],
    "app_hash": ""
  },
  "peers": [
    "cc30db6d116ca6ff440277c8e13e3ca253a774e7@sen1.testnet.leapdao.org:46691",
    "23d57231b620575c03f0e136fc37b5ec5812bb35@val.testnet.leapdao.org:46691",
    "deb472911a7a0ff3867b39c6ffa375ddbb42db5d@sen2.testnet.leapdao.org:46691",
    "5d27949fb42efc829bdc663f914fe0c069a96f2d@54.72.111.55:46691"
  ],
  "p2pPort": 46691,
  "nodeId": "cc30db6d116ca6ff440277c8e13e3ca253a774e7",
  "flagHeights": {
    "spend_cond_stricter_rules": 100000,
    "spend_cond_new_bytecode": 41000
  }
}
```

File: src/flags/flagsFactory.js (hunk 1 above)
Comment (sunify) [reply src/flags/flagsFactory.js]: In `saveState` i see only `bridgeState.currentState.blockHeight` update. For flags i use `bridgeState.blockHeight` which updates on each block.

Comment (sunify) [reply src/flags/flagsFactory.js]: Am i miss something?

**Comment by sunify** [comment]
Apparently there is no point to write any integration tests. Unit tests should be enough

File: src/flags/flagsFactory.js (hunk 1 above)
Comment (troggy) [reply src/flags/flagsFactory.js]: no, all is great 👍

Confused with `this.currentState.blockHeight`

File: README.md [hunk 2]
```diff
@@ -79,6 +79,38 @@ Your local port `9999` forwards now to the remote host on address `127.0.0.1` an
...
 - `networkId` - network ID. Possible values: `1340` - Leap mainnet, `1341` - Leap testnet.
 - `peers` — array of peers
+- `flagHeights` — `flag → height` mapping to enable feature flags only on certain heights (see [feature flags section](#feature-flags))
+
+### Feature flags
+
+If you want to introduce breaking changes (it can break existing networks) in tx checks you should wrap these changes into condition with feature flag to avoid [resync problems](https://github.com/leapdao/leap-node/issues/334).
+
+```es6
+if (bridgeState.flags.tx_should_fail_on_zero_input) {
```
Comment (troggy) [inline README.md]: <3

File: src/flags/flags.test.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    bridgeState.blockHeight = 10;
+    expect(flags.test_flag).toBe(true);
+  });
+
+  it('should throw on unknown flag', () => {
+    const bridgeState = { blockHeight: 0 };
+    const flags = makeFlags(bridgeState);
+    expect(() => {
+      const value = flags.unknown_flag; // eslint-disable-line no-unused-vars
+    }).toThrow('Unknow feature flag unknown_flag');
```
Comment (troggy) [inline src/flags/flags.test.js]:
```suggestion
    }).toThrow('Unknown feature flag unknown_flag');
```

File: src/flags/flags.test.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  });
+
+  it('should throw on attempt to set a flag value', () => {
+    const bridgeState = { blockHeight: 0 };
+    const flags = makeFlags(bridgeState);
+    expect(() => {
+      flags.test_flag = true;
```
Comment (troggy) [inline src/flags/flags.test.js]: very awesome!

File: src/flags/index.js [hunk 5]
```diff
@@ -0,0 +1,5 @@
+const flagsFactory = require('./flagsFactory');
+
+const FLAGS = ['spend_cond_stricter_rules', 'spend_cond_new_bytecode'];
```
Comment (troggy) [inline src/flags/index.js]:
```suggestion
// when adding a flag, add a link to the issue/PR describing the reason for the change
const FLAGS = [
  'spend_cond_stricter_rules', // https://github.com/leapdao/leap-node/pull/303
  'spend_cond_new_bytecode' // https://github.com/leapdao/leap-node/pull/292
];
```
```suggestion
const FLAGS = ['spend_cond_stricter_rules', 'spend_cond_new_bytecode'];
```

Comment (troggy) [reply src/flags/index.js]: poor examples, but still

File: src/tx/applyTx/checkSpendCond.js [hunk 6]
```diff
@@ -262,12 +262,20 @@ module.exports = async (state, tx, bridgeState, nodeConfig = {}) => {
...
-          );
+      if (input.signer && unspent.address !== input.signer) {
+        throw new Error(
+          `output owner ${unspent.address} unequal input signer: ${input.signer}`
+        );
+      }
+
+      // will throw if allowance is undefined
+      // stricter rule was introduced here
+      // https://github.com/leapdao/leap-node/blob/b534718807214318acaa6a9ff88b9cb8f1780ef1/src/tx/applyTx/checkSpendCond.js#L271
```
Comment (troggy) [inline src/tx/applyTx/checkSpendCond.js]: ❤️

File: src/flags/index.js (hunk 5 above)
Comment (sunify) [reply src/flags/index.js]: Awesome idea

File: src/flags/flags.test.js (hunk 3 above)
Comment (sunify) [reply src/flags/flags.test.js]: This typo was in module as well

**Pull request merged**